native/dlaqr4.go

```diff
@@ -0,0 +1,467 @@
+				wi[ks+1] = 0
+				wr[ks] = wr[ks+1]
+				wi[ks] = wi[ks+1]
+			}
+		} else {
+			// If we got ns/2 or fewer shifts, use Dlahqr on a
+			// trailing principal submatrix to get more. Since
+			// ns <= nsmax <=(n+6)/9, there is enough space below
+			// the subdiagonal to fit an ns×ns scratch array.
+			if kbot-ks+1 <= ns/2 {
+				ks = kbot - ns + 1
+				kt = n - ns
+				impl.Dlacpy(blas.All, ns, ns, h[ks*ldh+ks:], ldh, h[kt*ldh:], ldh)
+				inf := impl.Dlahqr(false, false, ns, 0, ns-1, h[kt*ldh:], ldh,
+					wr[ks:ks+ns], wi[ks:ks+ns], 0, 0, nil, 0)
+				ks += inf
+				// In case of a rare QR failure use eigenvalues
+				// of the trailing 2×2 principal submatrix.
+				if ks >= kbot {
+					aa := h[(kbot-1)*ldh+kbot-1]
+					bb := h[(kbot-1)*ldh+kbot]
+					cc := h[kbot*ldh+kbot-1]
+					dd := h[kbot*ldh+kbot]
+					_, _, _, _, wr[kbot-1], wi[kbot-1], wr[kbot], wi[kbot], _, _ =
+						impl.Dlanv2(aa, bb, cc, dd)
+					ks = kbot - 1
+				}
+			}
+
+			if kbot-ks+1 > ns {
+				// Sorting the shifts helps a little. Bubble
+				// sort keeps complex conjugate pairs together.
+				sorted := false
+				for k := kbot; k > ks; k-- {
+					if sorted {
+						break
+					}
+					sorted = true
+					for i := ks; i < k; i++ {
+						if math.Abs(wr[i])+math.Abs(wi[i]) >= math.Abs(wr[i+1])+math.Abs(wi[i+1]) {
+							continue
+						}
+						sorted = false
+						wr[i], wr[i+1] = wr[i+1], wr[i]
+						wi[i], wi[i+1] = wi[i+1], wi[i]
+					}
+				}
+			}
+
+			// Shuffle shifts into pairs of real shifts and pairs of
+			// complex conjugate shifts using the fact that complex
+			// conjugate shifts are already adjacent to one another.
+			// TODO(vladimir-ch): Dlaqr5 also does this shuffling.
+			// It seems that the one in Dlaqr5 is redundant but
+			// check.
+			for i := kbot; i > ks+1; i -= 2 {
+				if wi[i] == -wi[i-1] {
+					continue
+				}
+				wr[i], wr[i-1], wr[i-2] = wr[i-1], wr[i-2], wr[i]
+				wi[i], wi[i-1], wi[i-2] = wi[i-1], wi[i-2], wi[i]
+			}
+		}
+
+		// If there are only two shifts and both are real, then use only one.
+		if kbot-ks+1 == 2 && wi[kbot] == 0 {
+			if math.Abs(wr[kbot]-h[kbot*ldh+kbot]) < math.Abs(wr[kbot-1]-h[kbot*ldh+kbot]) {
+				wr[kbot-1] = wr[kbot]
+			} else {
+				wr[kbot] = wr[kbot-1]
+			}
+		}
+
+		// Use up to ns of the the smallest magnitude shifts. If there
+		// aren't ns shifts available, then use them all, possibly
+		// dropping one to make the number of shifts even.
+		ns = min(ns, kbot-ks+1)
+		ns -= ns % 2
+		ks = kbot - ns + 1
+
+		// Split workspace under the subdiagonal into:
+		// - a kdu×kdu work array U in the lower left-hand-corner,
+		// - a kdu×nho horizontal work array WH along the bottom edge
+		//   (nho must be at least kdu but more is better),
+		// - an nve×kdu vertical work array WV along the left-hand-edge
+		//   (nve must be at least kdu but more is better).
+		kdu := 3*ns - 3
+		ku := n - kdu
+		kwh := kdu
+		nho = n - (kdu + 3) - kdu
+		kwv = kdu + 3
+		nve = n - kdu - kwv
```

Comment:
You are not, I was also perplexed by this. I think they are trying to be very explicit and more general than necessary. H is square, U is square, so WH will inevitably has the same number of columns as WV rows. I merged nve and nho into nhv (also above before calling Dlaqr2).